From 39ee1064562bbf5a8443f9c90fb6f088c34f3785 Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Fri, 9 Oct 2020 19:40:59 +0530
Subject: [PATCH 655/767] octeontx2-bphy-netdev: Fix kernel crash with ioctl
 trying to add timer

This patch fixes the issue with ioctl OTX2_RFOE_IOCTL_PTP_OFFSET
which is trying to add the same instance of the ptp timer to the
timer list. When the ioctl is called multiple times, kernel crash
is observed.

Change-Id: I8ae760b9e2ff56a584e50fa194eb62e92e6e204c
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/37713
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c    | 6 +++---
 drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c     | 6 +++---
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
index 2fe1f3f97803..1691933c9d4f 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
@@ -232,6 +232,7 @@ static long otx2_bphy_cdev_ioctl(struct file *filp, unsigned int cmd,
 		struct ptp_clk_cfg clk_cfg;
 		struct net_device *netdev;
 		struct ptp_bcn_ref ref;
+		unsigned long expires;
 		int idx;
 
 		if (!cdev->odp_intf_cfg) {
@@ -286,9 +287,8 @@ static long otx2_bphy_cdev_ioctl(struct file *filp, unsigned int cmd,
 		memcpy(&ptp_cfg->old_ref, &ref, sizeof(struct ptp_bcn_ref));
 		memcpy(&ptp_cfg->new_ref, &ref, sizeof(struct ptp_bcn_ref));
 		ptp_cfg->use_ptp_alg = 1;
-		ptp_cfg->ptp_timer.expires = jiffies +
-					     PTP_OFF_RESAMPLE_THRESH * HZ;
-		add_timer(&ptp_cfg->ptp_timer);
+		expires = jiffies + PTP_OFF_RESAMPLE_THRESH * HZ;
+		mod_timer(&ptp_cfg->ptp_timer, expires);
 		ret = 0;
 		goto out;
 	}
diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
index 2cafee783505..e334aca7ca05 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
@@ -188,7 +188,7 @@ static void otx2_rfoe_ptp_offset_timer(struct timer_list *t)
 	struct ptp_bcn_off_cfg *ptp_cfg = from_timer(ptp_cfg, t, ptp_timer);
 	u64 mio_ptp_ts, ptp_ts_diff, ptp_diff_nsec, ptp_diff_psec;
 	struct ptp_clk_cfg *clk_cfg = &ptp_cfg->clk_cfg;
-	unsigned long flags;
+	unsigned long expires, flags;
 
 	spin_lock_irqsave(&ptp_cfg->lock, flags);
 
@@ -207,8 +207,8 @@ static void otx2_rfoe_ptp_offset_timer(struct timer_list *t)
 
 	spin_unlock_irqrestore(&ptp_cfg->lock, flags);
 
-	ptp_cfg->ptp_timer.expires = jiffies + PTP_OFF_RESAMPLE_THRESH * HZ;
-	add_timer(&ptp_cfg->ptp_timer);
+	expires = jiffies + PTP_OFF_RESAMPLE_THRESH * HZ;
+	mod_timer(&ptp_cfg->ptp_timer, expires);
 }
 
 /* submit pending ptp tx requests */
-- 
2.31.1

